File: crates/outl-frontend-shared/src/markdown/quote.ts
/**
 * Blockquote prefix helper — TS mirror of `outl_actions::quote`.
 *
 * Same wire-format policy as TODO/DONE: the marker lives **in the
 * block's text** as a `"> "` prefix (one space after `>`, matching
 * CommonMark) instead of a new field on the AST. Every client decides
 * how to render it.
 *
 * The Rust side owns the canonical implementation
 * (`crates/outl-actions/src/quote.rs`). This file exists so a client
 * that only has `OutlineNode.text` (mobile's `<BlockRow />`, desktop's
 * row chrome) doesn't need a Tauri round-trip to know if a row should
 * paint with a left border + muted body.
 */

/** Wire prefix for a quoted block — keep in sync with `QUOTE_PREFIX` in Rust. */
export const QUOTE_PREFIX = "> ";

/**
 * Split a block's raw text into `(quoted, body)`.
 *
 * When the prefix is present, `quoted` is `true` and `body` is
 * everything after `"> "` (no leading space). When absent, `quoted`
 * is `false` and `body` is the untouched input.
 *
 * The Rust mirror returns the same shape via `outl_actions::quote::split_quote`.
 */
export function splitQuote(text: string): { quoted: boolean; body: string } {
  if (text.startsWith(QUOTE_PREFIX)) {
    return { quoted: true, body: text.slice(QUOTE_PREFIX.length) };
  }
  return { quoted: false, body: text };
}

/** Does the block's raw text carry the quote marker? */
export function isQuote(text: string): boolean {
  return text.startsWith(QUOTE_PREFIX);
}

import type { InlineToken } from "../api/types";

/**
 * Strip the leading `"> "` prefix from the **first** `Plain` token of
 * a tokenized block body.
 *
 * Why this exists: the backend tokenizes the *raw* block text, so a
 * quoted block emits a single `Plain { value: "> hello" }` token (the
 * quote wrapper is block-level, not inline — see
 * `outl_md::inline::tokenize`). The mobile and desktop frontends
 * render the quote chrome themselves (left border + muted body) and
 * shouldn't double-paint the marker inside the rendered text. This
 * helper produces the token list to feed `<MarkdownInline />` when
 * the chrome already shows "this block is a quote".
 *
 * A no-op when the first token isn't a `Plain` starting with `"> "`
 * — keeps the helper safe to call unconditionally on every row.
 */
export function stripQuoteFromTokens(tokens: InlineToken[]): InlineToken[] {
  if (tokens.length === 0) return tokens;
  const first = tokens[0];
  if (first.kind === "plain" && first.value.startsWith(QUOTE_PREFIX)) {
    return [
      { ...first, value: first.value.slice(QUOTE_PREFIX.length) },
      ...tokens.slice(1),
    ];
  }
  return tokens;
}

File: crates/outl-actions/src/quote.rs
//! Quote state, encoded as a prefix on a block's text.
//!
//! Mirrors the TODO/DONE convention (see [`crate::todo`]): the marker
//! lives **in the block's text** as a `"> "` prefix instead of a new
//! field on the AST. Consumers (TUI, mobile, desktop) decide how to
//! render it; the wire format stays the same and round-trips through
//! CommonMark `.md` as `> body` continuation lines.
//!
//! Rules:
//!
//! 1. A block is a quote when its text starts with `"> "` — one space
//!    after `>`, matching CommonMark.
//! 2. Quote state is **per block**: children of a quoted block are not
//!    implicitly quoted (same policy as TODO/DONE).
//! 3. Nested quotes (`">> "`) are out of scope for the first cut — the
//!    body is whatever follows the single `"> "` prefix, verbatim, so a
//!    user who types `"> > foo"` gets a quoted block whose body
//!    *starts* with `"> foo"` but no extra unwrapping happens.

/// Wire prefix for a quoted block. Two characters including the
/// trailing space; consumers can rely on `chars().count() == 2` for
/// cursor math.
pub const QUOTE_PREFIX: &str = "> ";

/// Does the block's raw text carry the quote marker?
pub fn is_quote(text: &str) -> bool {
    text.starts_with(QUOTE_PREFIX)
}

/// Split a block's raw text into `(quoted, body)`. The body never
/// includes the prefix or its trailing space when `quoted` is true.
/// When the prefix is absent, `quoted` is `false` and `body` is the
/// untouched input.
pub fn split_quote(raw: &str) -> (bool, &str) {
    match raw.strip_prefix(QUOTE_PREFIX) {
        Some(rest) => (true, rest),
        None => (false, raw),
    }
}

/// Toggle the quote prefix on a block's raw text. Adds `"> "` when
/// absent, removes it when present. Mirrors
/// [`crate::todo::cycle_todo`] but with a two-state cycle (the quote
/// marker is binary, unlike TODO ↔ DONE ↔ none).
pub fn toggle_quote(raw: &str) -> String {
    let (quoted, body) = split_quote(raw);
    if quoted {
        body.to_string()
    } else {
        format!("{QUOTE_PREFIX}{raw}")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn split_recognises_the_marker() {
        assert_eq!(split_quote("> a quote"), (true, "a quote"));
        assert_eq!(split_quote("plain block"), (false, "plain block"));
    }

    #[test]
    fn marker_requires_trailing_space() {
        // ">foo" without the space stays plain (CommonMark requires
        // the space too).
        assert_eq!(split_quote(">foo"), (false, ">foo"));
        // A bare ">" is also not a quote (no body, no trailing space).
        assert_eq!(split_quote(">"), (false, ">"));
        assert!(!is_quote(">foo"));
        assert!(is_quote("> "));
    }

    #[test]
    fn toggle_walks_through_two_states() {
        let s0 = "ship the feature";
        let s1 = toggle_quote(s0);
        let s2 = toggle_quote(&s1);
        assert_eq!(s1, "> ship the feature");
        assert_eq!(s2, "ship the feature");
    }

    #[test]
    fn toggle_preserves_inner_text_verbatim_including_inline_tokens() {
        // Inline tokens inside a quote stay untouched — the wrapper
        // is transparent to the inline tokenizer.
        let s = toggle_quote("**bold** [[ref]] #tag");
        assert_eq!(s, "> **bold** [[ref]] #tag");

        let (quoted, body) = split_quote(&s);
        assert!(quoted);
        assert_eq!(body, "**bold** [[ref]] #tag");
    }

    #[test]
    fn empty_body_is_legal() {
        // "> " alone is a quoted block with an empty body — we treat
        // it as legal so the user can type the marker and then the
        // body without an intermediate "not a quote" state.
        assert_eq!(split_quote("> "), (true, ""));
        assert!(is_quote("> "));
    }
}
